scijava-ops-benchmarks/src/main/java/org/scijava/ops/benchmarks/Benchmarks.java

```diff
@@ -0,0 +1,189 @@
+
+/**
+ * Tests relative performance of ImageJ Ops and SciJava Ops.
+ *
+ * @author Gabriel Selzer
+ * @author Curtis Rueden
+ */
+@Fork(value = 1, warmups = 1)
+@Warmup(iterations = 2)
+@BenchmarkMode(Mode.AverageTime)
+@OutputTimeUnit(TimeUnit.MICROSECONDS)
+public class Benchmarks {
+
+	private static final Hints NO_CACHE = new Hints("cache.IGNORE");
+
+	// -- Benchmarks -- //
+
+	/**
+	 * Baseline benchmark: static code execution without Ops framework.
+	 *
+	 * @param state the state between benchmarks
+	 */
+	@Benchmark
+	public void noOps(BenchmarkState state) {
+		incrementByte(state.theByte);
+	}
+
+	/**
+	 * Static code execution (no Ops framework), adapting inplace operation to a
+	 * function.
+	 *
+	 * @param state the state between benchmarks
+	 */
+	@Benchmark
+	public void noOpsAdapted(BenchmarkState state) {
+		// Copy input data into result object, which will be mutated.
+		byte[] result = { state.theByte[0] };
+
+		// Mutate the result.
+		incrementByte(result);
+
+		// Store the result into the designated receptacle.
+		state.byteResult = result;
+	}
+
+	/**
+	 * Static code execution (no Ops framework), converting input data between
+	 * double and byte.
+	 *
+	 * @param state the state between benchmarks
+	 */
+	@Benchmark
+	public void noOpsConverted(BenchmarkState state) {
+		// Convert double value to byte.
+		byte[] convertedByte = toByte(state.theDouble);
+
+		// Call static method on converted byte.
+		incrementByte(convertedByte);
+
+		// Convert byte back to double.
+		double[] reconvertedDouble = toDouble(convertedByte);
+
+		// Copy result back to the original double receptacle.
+		state.theDouble[0] = reconvertedDouble[0];
+	}
+
+	/**
+	 * Static code execution (no Ops framework), doing both adaptation as
+	 * {@link #noOpsAdapted} and conversion as {@link #noOpsConverted}.
+	 *
+	 * @param state the state between benchmarks
+	 */
+	@Benchmark
+	public void noOpsAdaptedAndConverted(BenchmarkState state) {
+		// Convert double value to byte.
+		byte[] convertedByte = toByte(state.theDouble);
+
+		// Mutate the converted object. We can do this directly because it is a
+		// copy, not a wrapper/view.
+		incrementByte(convertedByte);
+
+		// Convert byte back to double, and store the result into the designated
+		// receptacle.
+		state.doubleResult = toDouble(convertedByte);
+	}
+
+	/**
+	 * ImageJ Ops.
+	 *
+	 * @param state the state between benchmarks
+	 */
+	@Benchmark
+	public void ijOps(final BenchmarkState state) {
+		state.ops.run("benchmark.increment", state.theByte);
+	}
+
+	/**
+	 * SciJava Ops, default behavior (caching included).
+	 *
+	 * @param state the state between benchmarks
+	 */
+	@Benchmark
+	public void sjOps(final BenchmarkState state) {
+		state.env.unary("benchmark.increment").input(state.theByte).mutate();
+	}
+
+	/**
+	 * SciJava Ops, caching disabled.
+	 *
+	 * @param state the state between benchmarks
+	 */
+	@Benchmark
+	public void sjOpsNoCache(final BenchmarkState state) {
```

Comment:
It's a little strange that this benchmark and the one below all ignore the cache, but only this one has `NoCache` in the name. But it's probably fine.

Reply:
I know, I don't like it. I was going to rename it to be more consistent, but forgot. I'll take care of it!